include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,665 @@
+     * which will use either KINSOL or NOX depending on which package is
+     * installed, KINSOL having priority.
+     */
+    SolverType solver_type;
+
+    /**
+     * The solution strategy to use. For this class, you can choose from
+     * SolutionStrategy::newton, SolutionStrategy::linesearch, or
+     * SolutionStrategy::picard. More details on this can be found on the
+     * @p KINSOL documentation.
+     */
+    SolutionStrategy strategy;
+
+    /**
+     * Maximum number of nonlinear iterations allowed.
+     */
+    unsigned int maximum_non_linear_iterations;
+
+    /**
+     * A scalar used as a stopping tolerance on the scaled
+     * maximum norm of the system function $F(u)$ or $G(u)$.
+     *
+     * If set to zero, default values provided by KINSOL will be used.
+     */
+    double function_tolerance;
+
+    /**
+     * Relative $l_2$ tolerance of the residual to be reached.
+     *
+     * @note Solver terminates successfully if either the absolute or
+     * the relative tolerance has been reached.
+     */
+    const double relative_tolerance;
+
+    /**
+     * A scalar used as a stopping tolerance on the minimum
+     * scaled step length.
+     *
+     * If set to zero, default values provided by KINSOL will be used.
+     */
+    double step_tolerance;
+
+    /**
+     * The size of the subspace used with Anderson acceleration
+     * in conjunction with Picard or fixed-point iteration.
+     *
+     * If you set this to 0, no acceleration is used.
+     */
+    unsigned int anderson_subspace_size;
+  };
+
+  /**
+   * Constructor, filling in default values
+   */
+  NonlinearSolverSelector();
+
+  /**
+   * Constructor, selecting the solver and other parametersspecified in
+   * @p additional_data.
+   */
+  NonlinearSolverSelector(const AdditionalData &additional_data);
+
+  /**
+   * Constructor
+   *
+   * @param additional_data NonlinearSolverSelector configuration data
+   * @param mpi_communicator MPI communicator over which logging operations are
+   * computer.
+   */
+  NonlinearSolverSelector(const AdditionalData &additional_data,
+                          const MPI_Comm &      mpi_communicator);
+
+  /**
+   * Select a new nonlinear solver. All solver names used in this class are
+   * all lower case.
+   */
+  void
+  select(const typename AdditionalData::SolverType &type);
+
+/**
+ * Set the additional data. For more information see the @p Solver class.
+ */
+#ifdef DEAL_II_WITH_TRILINOS
+  void
+  set_data(
+    const typename TrilinosWrappers::NOXSolver<VectorType>::AdditionalData
+      &                                         additional_data,
+    const Teuchos::RCP<Teuchos::ParameterList> &parameters =
+      Teuchos::rcp(new Teuchos::ParameterList));
+#endif
+
+/**
+ * Set the additional data. For more information see the @p Solver class.
+ */
+#ifdef DEAL_II_WITH_SUNDIALS
+  void
+  set_data(const typename SUNDIALS::KINSOL<VectorType>::AdditionalData
+             &additional_data);
+#endif
```

Comment:
If you call these functions, does this information have precedence over the `AdditionalData` information? Or the other way around? This should be documented.